core/src/executor/fetch.rs

```diff
@@ -150,12 +149,26 @@ pub async fn fetch_relation_rows<'a, T: GStore>(
 
                         Rows::Indexed(rows)
                     }
-                    Some(IndexItem::PrimaryKey(expr)) => {
-                        let filter_context = filter_context.as_ref().map(Rc::clone);
-                        let key = evaluate(storage, filter_context, None, expr)
-                            .await
-                            .and_then(Value::try_from)
-                            .and_then(Key::try_from)?;
+                    Some(IndexItem::PrimaryKey(primary_key_expressions)) => {
+                        let mut evaluated_primary_key_expressions = Vec::new();
+
+                        for expr in primary_key_expressions {
+                            evaluated_primary_key_expressions.push(Value::try_from(
+                                evaluate(
+                                    storage,
+                                    filter_context.as_ref().map(Rc::clone),
+                                    None,
+                                    expr,
+                                )
+                                .await?,
+                            )?);
+                        }
+
+                        let key = if evaluated_primary_key_expressions.len() == 1 {
+                            Key::try_from(evaluated_primary_key_expressions.pop().unwrap())?
+                        } else {
+                            Key::try_from(evaluated_primary_key_expressions)?
+                        };
```

Comment:
nit. using stream crate is quite useful here, like this below.

```rust
Some(IndexItem::PrimaryKey(exprs)) => {
    let key = stream::iter(exprs)
        .then(move |expr| {
            let filter_context = filter_context.as_ref().map(Rc::clone);

            async move {
                evaluate(storage, filter_context, None, expr)
                    .await
                    .and_then(Value::try_from)
                    .and_then(Key::try_from)
            }
        })
        .try_collect::<Vec<Key>>()
        .await
        .and_then(|keys| match keys.len() {
            1 => keys
                .into_iter()
                .next()
                .ok_or(FetchError::UnreachableEmptyPrimaryKey.into()),
            _ => Ok(Key::List(keys)),
        })?;
```

and we can consider adding an unreachable prefixed error to FetchError. 
```rust
    #[error("unreachable - primary key vector is empty")]
    UnreachableEmptyPrimaryKey,
```

Reply:
Refactored fetch code as described.